### PR TITLE
feat: [FC-0047] Full-Bleed Header + Top Navigation

### PR DESCRIPTION
**Pull request by IvanStepanok**
This PR introduces updated navigation and banner for the course home.

Light and Dark theme:
Simulator Screen Recording - iPhone 15 Pro Max - 2024-04-02 at 17 06 12 Simulator Screen Recording - iPhone 15 Pro Max - 2024-04-02 at 17 15 01

In the horizontal state, the top bar is always collapsed.
Simulator Screen Recording - iPhone 15 Pro Max - 2024-04-02 at 17 19 22

The iPad version doesn't have a collapsing effect.
Simulator Screen Recording - iPad (10th generation) - 2024-04-02 at 17 22 35

:rotating_light: **Breaking changes** :rotating_light::
The `COURSE_BANNER_ENABLED` and `COURSE_TOP_TAB_BAR_ENABLED` feature flags have been removed.

### CODE REVIEW

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @IvanStepanok! Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


This is currently a draft pull request. When it is ready for our review and all tests are green, click "Ready for Review", or remove "WIP" from the title, as appropriate.

**Review by marcotuts — APPROVED** [review]
Product review only 👍

**Review by volodymyr-chekyrta — DISMISSED** [review]

**Comment by rnr** [comment]
Issues with loading indicator:
1) Changed loading indicator for Course Home view (is it approved?):

Before|After
--|--
|

2) Loading indicator for "More" tab:

**Comment by rnr** [comment]
Issues with Scroll
1) When user scrolls on Discussion tab it doesn't collapse header (how it does on Home tab for example)

https://github.com/openedx/openedx-app-ios/assets/37253/3d66d306-396e-431c-9dbf-d4d07ea53166

2) Can't scroll to bottom on Dates tab (I guess it happens when content takes up a bit more space than is available on the screen). Strange animation for this case:

https://github.com/openedx/openedx-app-ios/assets/37253/74a1b81f-6d77-4226-bfdb-264da01fbcdf

**Comment by saeedbashir** [comment]
The progress bar location is odd if we use pull to refresh while data is loading. and If we interact with UI white loading the bleeding header isn't working properly. 
https://github.com/openedx/openedx-app-ios/assets/5606473/161d1427-be69-420f-858a-10a14341aca7

**Comment by IvanStepanok** [comment]
## @rnr Thanks for your review!🙌
1. Old loading indicators are returned. 
Simulator Screenshot - iPhone 15 Pro Max - 2024-04-18 at 13 44 02

2. Fixed the “More” loading indicator.
Simulator Screenshot - iPhone 15 Pro Max - 2024-04-18 at 13 45 19

3. Pull to refresh indicators are equal now.
Simulator Screenshot - iPhone 15 Pro Max - 2024-04-18 at 13 45 48 Simulator Screenshot - iPhone 15 Pro Max - 2024-04-18 at 13 45 42

## Scroll issues
1. The discussion tab collapsing issue is fixed.

https://github.com/openedx/openedx-app-ios/assets/128456094/d34c4464-da9c-4d95-b08f-df2efec52c52

2. The animation glitch on some scroll views is fixed by adding new logic.

https://github.com/openedx/openedx-app-ios/assets/128456094/7666636d-e581-4fe7-81b5-15681eca1321

If the content height in ScrollView is not enough to collapse the header, collapsing mode are disabled. Thanks for paying attention to this bug, it's very important.🙌

## @saeedbashir Thanks for your review, too!❤️

1. Fixed a bug with changes in header size after loading data.
2. The header is no longer transparent during loading.
3. Progress bars are currently in the normal position.

https://github.com/openedx/openedx-app-ios/assets/128456094/b69909f2-198f-48da-8da0-7554da32484b

If you discover any new bugs, I'd be delighted to fix them!

**Comment by rnr** [comment]
> If the content height in ScrollView is not enough to collapse the header, collapsing mode are disabled.
I think we need to disable collapse mode for the 'More' tab as it tries to collapse, but we have two rows (Distributions and Announcements) at the moment. Thank you

https://github.com/openedx/openedx-app-ios/assets/37253/a10624a8-7030-4582-8b95-349083877d4f

**Comment by rnr** [comment]
Is it a design decision to disable this collapse mode for iPad in general?


https://github.com/openedx/openedx-app-ios/assets/37253/b480a039-e9c7-417f-b9e1-a9514c94b27f

File: Course/Course/Presentation/Subviews/TopHeaderView.swift [hunk 1]
```diff
@@ -0,0 +1,173 @@
+//
+//  TopHeaderView.swift
```
Comment (saeedbashir) [inline Course/Course/Presentation/Subviews/TopHeaderView.swift:2]: How about giving it a more meaningful name like course header view or some as per its functionality Thoughts?

File: Course/Course/Presentation/Subviews/TopHeaderView.swift [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+    var title: String
+    @Binding var collapsed: Bool
+    var containerWidth: CGFloat
+    var animationNamespace: Namespace.ID
+    @Binding private var isAnimatingForTap: Bool
+    @Environment(\.isHorizontal) private var isHorizontal
+    
+    let collapsedHorizontalHeight: CGFloat = 230
+    let collapsedVerticalHeight: CGFloat = 260
+    let expandedHeight: CGFloat = 300
```
Comment (saeedbashir) [inline Course/Course/Presentation/Subviews/TopHeaderView.swift:25]: All these variables can be private.

File: Course/Course/Presentation/Subviews/TopHeaderView.swift [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+    
+    private struct GeometryName {
+        static let backButton = "backButton"
+        static let topTabBar = "topTabBar"
+        static let blurSecondaryBg = "blurSecondaryBg"
+        static let blurPrimaryBg = "blurPrimaryBg"
+        static let blurBg = "blurBg"
+    }
```
Comment (saeedbashir) [inline Course/Course/Presentation/Subviews/TopHeaderView.swift:33]: Do we need a struct here, how about converting it to an enum?

File: Course/Course/Presentation/Subviews/TopHeaderView.swift [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+                        .onFailureImage(CoreAssets.noCourseImage.image)
+                        .resizable()
+                        .aspectRatio(contentMode: .fill)
+                        .allowsHitTesting(false)
+                        .clipped()
+                        .background(Theme.Colors.background)
+                }
+                }
+                .disabled(true)
+                .ignoresSafeArea()
```
Comment (saeedbashir) [inline Course/Course/Presentation/Subviews/TopHeaderView.swift:66]: The code needs formatting here.

File: Course/Course/Presentation/Subviews/TopHeaderView.swift [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+        .background(Theme.Colors.background)
+        .frame(
+            height: collapsed ? (
+                isHorizontal ? collapsedHorizontalHeight : collapsedVerticalHeight
+            ) : expandedHeight
+        )
+        .ignoresSafeArea(edges: .top)
+    }
+    
+    private func topTabBar(containerWidth: CGFloat) -> some View {
```
Comment (saeedbashir) [inline Course/Course/Presentation/Subviews/TopHeaderView.swift:160]: How about renaming it to courseMenuBar or courseTabbar? thoughts?

File: Core/Core/View/Base/RefreshProgressView.swift [hunk 6]
```diff
@@ -0,0 +1,22 @@
+//
+//  RefreshProgressView.swift
+//  Course
+//
+//  Created by  Stepanok Ivan on 15.04.2024.
+//
+
+import SwiftUI
+
+public struct RefreshProgressView: View {
```
Comment (saeedbashir) [inline Core/Core/View/Base/RefreshProgressView.swift:10]: I guess it's being used for pull to refresh, how about renaming it as per functionality thoughts?

File: Core/Core/View/Base/ResponsiveView.swift [hunk 7]
```diff
@@ -0,0 +1,70 @@
+//
+//  ResponsiveView.swift
+//  Core
+//
+//  Created by  Stepanok Ivan on 26.03.2024.
+//
+
+import SwiftUI
+
+public struct ResponsiveView: View {
```
Comment (saeedbashir) [inline Core/Core/View/Base/ResponsiveView.swift:10]: Is it possible to give it a meaningful name as every view is by default responsive view.

File: Core/Core/View/Base/ResponsiveView.swift [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+    
+    private let padHeight: CGFloat = 290
+    private let collapsedHorizontalHeight: CGFloat = 120
+    private let collapsedVerticalHeight: CGFloat = 100
+    private let expandedHeight: CGFloat = 240
+    private let coordinateBoundaryLower: CGFloat = -115
+    
+    @State var lastHeight: CGFloat = 240
```
Comment (saeedbashir) [inline Core/Core/View/Base/ResponsiveView.swift:18]: It can be private

File: Core/Core/View/Base/ResponsiveView.swift [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+        VStack {
+        }.frame(
+            height: idiom == .pad
+            ? padHeight
+            : (collapsed ? (isHorizontal ? collapsedHorizontalHeight : collapsedVerticalHeight) : expandedHeight)
+        )
+        .overlay(
+            GeometryReader { geometry -> Color in
+                guard idiom != .pad, enableCollapsing else {
+                    return Color.clear
```
Comment (saeedbashir) [inline Core/Core/View/Base/ResponsiveView.swift:47]: Color is redundant, you can directly use .clear.

**Review by saeedbashir — CHANGES_REQUESTED** [review]
* The header is not collapsing on home and videos tab.
* The back button is not showing the menu on long press.

https://github.com/openedx/openedx-app-ios/assets/5606473/419ae958-b49a-4f7e-a447-7687c4507113

File: Core/Core/View/Base/ResponsiveView.swift (hunk 9 above)
Comment (IvanStepanok) [reply Core/Core/View/Base/ResponsiveView.swift:47]: Thanks! Fixed

File: Core/Core/View/Base/ResponsiveView.swift (hunk 8 above)
Comment (IvanStepanok) [reply Core/Core/View/Base/ResponsiveView.swift:18]: Thanks! Fixed

File: Core/Core/View/Base/ResponsiveView.swift (hunk 7 above)
Comment (IvanStepanok) [reply Core/Core/View/Base/ResponsiveView.swift:10]: Renamed to DynamicOffsetView. Good?

File: Course/Course/Presentation/Subviews/TopHeaderView.swift (hunk 3 above)
Comment (IvanStepanok) [reply Course/Course/Presentation/Subviews/TopHeaderView.swift:33]: Thanks! Good idea

File: Course/Course/Presentation/Subviews/TopHeaderView.swift (hunk 1 above)
Comment (IvanStepanok) [reply Course/Course/Presentation/Subviews/TopHeaderView.swift:2]: Renamed to CourseHeaderView

File: Course/Course/Presentation/Subviews/TopHeaderView.swift (hunk 2 above)
Comment (IvanStepanok) [reply Course/Course/Presentation/Subviews/TopHeaderView.swift:25]: Thanks! Fixed

File: Course/Course/Presentation/Subviews/TopHeaderView.swift (hunk 4 above)
Comment (IvanStepanok) [reply Course/Course/Presentation/Subviews/TopHeaderView.swift:66]: sure

**Comment by IvanStepanok** [comment]
https://github.com/openedx/openedx-app-ios/assets/128456094/2ed59cae-2fde-4290-adf6-3edcfd47cc53

@rnr, @saeedbashir, thanks for the review, I appreciate it. Feel free to reach out if you find any new bugs or have questions. Thanks again!

**Review by saeedbashir — DISMISSED** [review]
LGTM 👍

**Comment by IvanStepanok** [comment]
> Is it a design decision to disable this collapse mode for iPad in general?
> 
>  Screen.Recording.2024-04-19.at.11.37.37.mov

Yes, it is approved with @sdaitzman

**Comment by forgotvas** [comment]
Header bar hides logo by half or more on iPad:


iPhone| iPad Portrait| iPad Landscape
--|--|--
Simulator Screenshot - iPhone 15 - 2024-04-22 at 12 36 26|big1|big2

**Comment by forgotvas** [comment]
Back button in wrong position for pages with header on iPad:

Course home| Course home (Landscape) | Content | Content (Landscape)
--|--|--|--
Simulator Screenshot - iPad (10th generation) - 2024-04-22 at 12 39 01|Simulator Screenshot - iPad (10th generation) - 2024-04-22 at 12 39 34|Simulator Screenshot - iPad (10th generation) - 2024-04-22 at 12 39 03|Simulator Screenshot - iPad (10th generation) - 2024-04-22 at 12 39 31

**Comment by forgotvas** [comment]
Controller with header doesn't have navigation title:

https://github.com/openedx/openedx-app-ios/assets/480059/26971cc7-bc31-4a92-b997-4b9c1c434bef

**Comment by forgotvas** [comment]
Resume button doesn't have paddings on top:

develop | FC-0047
--|--
Simulator Screenshot - iPad (10th generation) - 2024-04-22 at 12 46 30|Simulator Screenshot - iPad (10th generation) - 2024-04-22 at 12 46 32

**Comment by forgotvas** [comment]
Pull to refresh indicator disappeared from videos tab:

Develop:
https://github.com/openedx/openedx-app-ios/assets/480059/4ec46061-f710-4ca4-9e59-31c51a35a5a9

FC-0047:
https://github.com/openedx/openedx-app-ios/assets/480059/21c9e8b9-091a-4179-beb1-022c0661eac6

File: Course/Course/Presentation/Outline/ContinueWithView.swift [hunk 10]
```diff
@@ -40,7 +40,6 @@ struct ContinueWithView: View {
                         UnitButtonView(type: .continueLesson, action: action)
                             .frame(width: 200)
                     } .padding(.horizontal, 24)
-                        .padding(.top, 32)
```
Comment (forgotvas) [inline Course/Course/Presentation/Outline/ContinueWithView.swift:43]: Are you sure that we don't need that?

**Comment by forgotvas** [comment]
Thanks, @IvanStepanok, just one was skipped - back button has wrong position it should be at same position as on content page: 

course home | content
--|--
Simulator Screenshot - iPad (10th generation) - 2024-04-22 at 22 35 47|Simulator Screenshot - iPad (10th generation) - 2024-04-22 at 22 42 11

**Review by forgotvas — APPROVED** [review]
LGTM

**Pull request merged**